### PR TITLE
[ci] update CI stuff

### PR DESCRIPTION
**Pull request by StrikerRUS**
- Use our Ubuntu 14.04 Docker for GPU test on Azure Pipelines.
- Update CMake and Clang in Docker to new stable versions.
- Install `libomp-dev` inside Docker instead of via the setup script for Clang.
- Update MinGW-w64 on Appveyor.
- Update Boost library.
- Update Boost submodule.

Also, this PR resolves the issue about GPU test described in https://github.com/Microsoft/LightGBM/pull/2078#issuecomment-479605231

### CODE REVIEW

**Comment by StrikerRUS** [comment]
Inverted compilers (Clang instead gcc and gcc instead Clang) work fine too:
3ecf8965ec66c7a62cf3f7d97e902c4f261e8c0d

image

File: .vsts-ci.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ variables:
 resources:
   containers:
   - container: ubuntu1404
-    image: lightgbm/vsts-agent:ubuntu-14.04
+    image: lightgbm/vsts-agent:ubuntu-14.04-dev
```
Comment (StrikerRUS) [inline .vsts-ci.yml]: This is to test and show that everything works.
I'll remove `dev` postfix before merging.

**Comment by guolinke** [comment]
@StrikerRUS this is to fix current fails of CI?

**Comment by StrikerRUS** [comment]
@guolinke Yeah

**Pull request merged**